### ISSUE TITLE
bpf: do atomic swapping of PolicyMap per endpoint

### ISSUE THREAD

**Issue by ianvernon** [cilium/cilium#3897]
When we update the policymap for an endpoint, we do it on a per-entry basis in the map that is currently live in the datapath. Instead, when regenerating policy, create a new map, and if all insertion into that map is successful, atomically swap the maps that are used for the endpoint in the datapath once policy recomputation is complete.

**Comment by joestringer** [comment]
This can be done using "map-in-map" (`BPF_MAP_TYPE_ARRAY_OF_MAPS` / `BPF_MAP_TYPE_HASH_OF_MAPS`).

**Comment by ianvernon** [comment]
Hm, we do not have those in the file that we include:
https://github.com/cilium/cilium/blob/b814d8341e50beeadf6d004cb422eb7dce2f74d4/bpf/include/linux/bpf.h#L86-L99

But we have them in the golang counterpart of the above:

https://github.com/cilium/cilium/blob/b814d8341e50beeadf6d004cb422eb7dce2f74d4/pkg/bpf/bpf.go#L36-L52

Is this intentional? I am a BPF newbie so I could be completely wrong ;) 

cc: @borkmann

**Comment by joestringer** [comment]
@ianvernon We just haven't synced the headers for a while, and since we're not using these variables yet it doesn't matter. We'll need to pull in newer linux `include/uapi/linux/bpf.h` when implementing this feature.

**Comment by borkmann** [comment]
Note that this feature requires a kernel version of 4.12 or higher. Not a problem, but we need some compat wrappers and feature probe detection for 4.9 - 4.11 kernels.

**Comment by ianvernon** [comment]
Discussed offline with @joestringer ; we think that this will take a large amount of work given that the code freeze for 1.2 is quite soon. 

Some notes from the discussion:

* The probe should be fairly similar to what we have for LRU
* Given that there is datapath work, there are some variabilities w.r.t. time for completing this (hitting the complexity limit due to new map lookups, which will add ~10 new instructions).
* Work around map swapping; where would this be done? Userspace? 
* The outer map basically ends up with 2 entries, FDs for old and new maps, then the datapath needs to either index into [0] or [1] depending on which one is old and which is new.
* CLI tools would also need updating, to pick up the appropriate map
* CI only tests Linux 4.9 kernel, need a way to CI-test this

**Comment by ianvernon** [comment]
Removing the 1.2 milestone and deferring to 1.3.

**Comment by ianvernon** [comment]
Question - instead of using an array of maps, given that the number of policy maps is bounded for a given endpoint with this change (one that is in use, and the other that is ready to be atomically swapped), instead of using `BPF_MAP_TYPE_ARRAY_OF_MAPS` / `BPF_MAP_TYPE_HASH_OF_MAPS`, is there a way that we can perform some type of `mv` operation from some of our scripts to swap the map locations on disk? That way we wouldn't be restricted to using > 4.12 kernels. Forgive me if this is an ignorant question :)

**Comment by borkmann** [comment]
@ianvernon Yeah, so what you could do instead is to have a new map in the bpffs with a different name, load all the policy data into it, mv the old map to a temp location and mv the new map to the actual map name, then trigger an atomic reload of the bpf program, and it will pick the new map to be used. So instead of relying on atomic map in map replacement, you would use atomic prog replacement, that should be fine to do, yes.

**Comment by ianvernon** [comment]
@borkmann great, thanks! I'll see if that is possible. We should definitely have the old map around as well if compilation fails; I think what we should do is create two copies of the endpoint policy map on the bpffs that are identical (one which is used in the datapath, the other which is a copy of it that we can keep around in case we have to revert to it when say, a program compilation fails or the move operation fails). So, we'd have three maps for say, endpoint ID 17051:

* `cilium_policy_17051`
    * map that is used for policy lookup (realized state)
* `cilium_policy_17051_realized` (open to better names for this)
    * map that is identical to 17051 before swapping - if swap or next operations related to BPF program compilation failure occur, then we `mv cilium_policy_17051_realized cilium_policy_17051` . We could have a controller in userspace which always tries to match the contents of the "realized" PolicyMap state for the endpoint in userspace with this BPF map. 
    * If we `mv cilium_policy_17051_realized cilium_policy_17051`, then we have to repopulate it before we add it again. This could be part of a "recovery" step for a failed build for an endpoint.
* `cilium_policy_17051_desired`
    * map that contains to-be-used PolicyMap for this endpoint, which would be populated once policy is computed for an endpoint via a controller.

Thoughts?

**Comment by joestringer** [comment]
@ianvernon Do you mean to say that there are three possible map names, but only ever two maps - desired/pending and realized/active?

If the compilation/load fails, we don't have to touch the maps at all. The map switch from one map to the other should only occur when the BPF program is atomically swapped on the device. That is to say, we could implement it in the following way and have just the path that refers to the realized state, and the path that refers to the pending map:

1. Assume that a BPF program is active, using `cilium_policy_17051`.
1. A new policy is pushed into Cilium
1. Cilium calculates the new BPF map contents for the new policy
   1. Cilium populates the `/sys/fs/bpf/tc/globals/cilium_policy_17051_pending` map with the new contents
1. Cilium triggers the `bpf/join_ep.sh` script, which:
   1. Creates the `/var/run/cilium/state/17051/bpf_lxc.o` object
   1. Attempts to atomically swap the BPF program on the device.
      1. If this succeeds, move `cilium_policy_17051_pending` to `cilium_policy_17051`, replacing the existing (realized) map path
      1. If this fails, do nothing (the existing map will remain in operation; the next step will clean up the pending map path).
1. Assuming that `join_ep.sh` succeeds, the controller for BPF policy must reopen the standard BPF map path `cilium_policy_17051` for syncing.
   1. If this fails, we should delete the pending map or move it to `cilium_policy_17051_failed`.
      1. If we decide to keep it around at the failed path location, this should be deleted when the endpoint is deleted.

From a datapath perspective, (3.i) just requires the `POLICY_MAP` macro definition to refer to `cilium_policy_17051_pending` rather than `cilium_policy_17051`. Obviously from userspace the policy map population logic would need to open this pending path instead of the existing map path.

If following the above order, then each time we recalculate the new policy, we would create a new pending policy map in the kernel (3). Every time we successfully run `join_ep.sh` it would remove the last reference to the old policy map (4.ii.a) and hence trigger the kernel to clean it up. 

Note that we could alternatively retain two map instances and swap them all the time. However, this would require cleaning up the old map from userspace which could involve many syscalls, whereas if we create a new map fresh each time then that step would be unnecessary and we wouldn't need logic to delete old entries from the map. So I would expect that creating a new map each time would actually be fairly cheap if more than one or two elements in the map change each time BPF needs to be recompiled.

**Comment by ianvernon** [comment]
> The map switch from one map to the other should only occur when the BPF program is atomically swapped on the device.

Good point, this reduces the need for having to revert back - not sure why I didn't think of this in the first place. 

The rest of your proposal sounds good. It sounds like we have come to a consensus about how to get this done. I'll get started on this :D

**Comment by joestringer** [comment]
Note that this will mean that `join_ep.sh` needs to be run upon policy change, however it does not guarantee that we need to actually compile the BPF prog. These are currently tied together. Might be worth considering putting a little thought into that so as not to regress on performance for changing policy.

**Comment by ianvernon** [comment]
@joestringer good point - a flag to the script should be easy enough to make such considerations.

**Comment by ianvernon** [comment]
After analysis of the code, it looks like we still have to do some updating of the maps that are live  after compilation (atomically swapping using `mv <new map> <old map>`). To illustrate why, the current PolicyMap lifecycle is as follows:

* compute endpoint's `desiredPolicyMapState` during `regeneratePolicy`
* sync `PolicyMap` on BPFFS with `desiredPolicyMapState` - this is necessary to do before recompiling BPF because we want to remove from the map proxyports which will no longer be used (in the case of a rule going from L7 --> L4).
* recompile BPF program (if necessary)
* generate / add new redirects for proxy (this populates `desiredPolicyMapState` with the new redirects)
* remove old / obsolete proxy redirects (this can be done now since we updated the PolicyMap above)
* wait for proxy to implement new redirects (this *must* happen after BPF compilation, specifically because of the sidecar proxy case):
    * via @rlenglet : We need to install the BPF program to give L3/L4 connectivity to the sidecar proxy so it can connect to Pilot and get configured, so it can install the Cilium filters, which can consume and ACK the network policies. Otherwise, `WaitForProxyCompletions` would never succeed
* but, after we wait for the proxy to implement the set of redirects that we want, we have to update the PolicyMap again to contain these redirects.

This means that after we compile the program, we have to update the realized state while it is being used in the datapath (which is what this issue is trying to avoid, unfortunately). The reason this has to be done, is because, according to @borkmann and @joestringer , the BPF program for the endpoint uses the map at the path provided at the time of compilation - this means that if we update the desired PolicyMap state after compilation, and then swap out the maps on the filesystem, the BPF program is still pointing to the old map. So, unless we recompile the program again (which is a huge performance cost), we have no choice but to update the map again after compiling. We could potentially do the atomic swap before recompiling BPF, and then update the realized map state after the proxy has completed realizing the redirects. This still provides us the benefit of not updating the PolicyMap inline before the program has been compiled.

**Comment by joestringer** [comment]
Thanks for the analysis, makes sense. One minor adjustment note for clarity below.

> The reason this has to be done, is because, according to @borkmann and @joestringer , the BPF program for the endpoint uses the map at the path provided at the time of compilation - this means that if we update the desired PolicyMap state after compilation, and then swap out the maps on the filesystem, the BPF program is still pointing to the old map. So, unless we recompile the program again (which is a huge performance cost), we have no choice but to update the map again after compiling. 

At the time of compilation, the header file changes (including map paths) are baked into the ELF file. The BPF program uses the path in the ELF at ***load*** time. So, if you can avoid changing the path to the desired map, then you can avoid recompiling. However, you would still need to populate the map at the correct location and load the program into the kernel again.

**Comment by joestringer** [comment]
> * sync PolicyMap on BPFFS with desiredPolicyMapState - this is necessary to do before recompiling BPF because we want to remove from the map proxyports which will no longer be used (in the case of a rule going from L7 --> L4).

Does this push down L3/L4 entries that we will subsequently configure an L7 redirect for? Ie, if you have a policy that allows `GET /public` on `foo:80`, will it first allow at L3/L4 on port 80 to foo and then afterwards configure the redirect? It seems to me like such a case should actually hold off on pushing the policymap entry into the datapath until the redirect is configured, otherwise in the case where we go from deny-all -> allow at L7, we will have an intermediate step where more traffic is allowed than should be allowed. *(That said, if we previously allowed at L3/L4 and move to a policy that allows with L7 restrictions then this would briefly put us into deny during a policy transition, which could lose packets that should legitimately be allowed in both the previous policy and the new one. I'll go into a mitigation of this below)*

I think that performing atomic updates into the pending map for the bulk of the changes (including L3/L4 and any L7 redirect changes that don't require interaction with the proxy) first, then subsequently updating the realized map for new redirects (creating the L3+L4 policymap entry including redirect for each entry at once) seems like a reasonable step that would improve the rigidity of policy transitions above what we have right now.

To be explicit:

1. Configure the `..._pending` map with:
   * Any entries that will be allowed by L3
   * Any entries that will be allowed by L3+L4
   * Any entries that will be allowed by L3+L4+L7 with redirects that are known
     * If the redirect is not configured, this will be deferred to the last step.
   * Any entries that were previously allowed at L3+L4 and will become L3+L4+L7 with a redirect configured *(This prevents the drop during L4->L7 policy transition above)*
     * The redirect part of this is deferred to the last step
1. Compile *(If necessary)*, attach the new BPF program.
1. Generate, configure any new redirects into the proxy
1. Populate the realized map with:
   * Any entries that will be allowed by L3+L4+L7 *(that were previously in drop)*
   * Update any entries that were previously allowed at L3+L4 to include redirect

Strictly speaking, the last step could be performed as another creation of `...pending` map for all L3/L4/L7 policymap entries, then a re-attach of the BPF program, which would provide even stronger guarantees about the policy state. However, I suspect that this will be quite costly(*) and it's not immediately obvious that the added cost provides a strong benefit.

*: This means two BPF map update syscalls per L3/L4 entry and ~one per L7 entry, rather than today's one per L3/L4/L7 entry.

Thoughts?

**Comment by rlenglet** [comment]
@joestringer Here is what `regenerateBPF` does now:

1. Populate the policy map's desired state with (a) all the L3/L4 entries, (b) the L7 entries for redirects that  existed in the previous revision and still exist in the new revision. In effect, this invalidates any obsolete L7 redirect. And note that the new redirects are not defined yet. (c) Sync the map.
2. Push the updated policy to Envoy. But don't wait for an Envoy ACK yet.
3. Regenerate and install the BPF program.
4. (a) Configure the new redirects in Envoy. (b) Update the policy map's desired state with the new redirects (update or create new entries with non-zero proxy ports).
5. Un-configure the obsolete redirects in Envoy.
6. Wait for an ACK from Envoy.
7. Sync the policy map.

This already does the right thing. Nothing needs to change IMHO. In particular, this meets all the constraints:

- 2, 4(a) and 5 all update Envoy's configuration, and therefore must happen before 6 to be ACKed.
- 3 must happen before 6 so that a sidecar proxy has L3/L4 connectivity and can start and get configured, so that it can install the Cilium L7 filter, which can consume and ACK the configuration sent in 2, 4(a) and 5.
- 1(c) must happen before 5 so that we are sure we're not going to send traffic to obsolete redirects before we un-configure those redirects in Envoy.
- 4(a) + 6 must happen before 4(b) + 7 because we must be sure new redirects are configured in Envoy before we start sending traffic to it in the policy map.

We may choose to move a few things while still meeting those constraints:
- 4(b) can be done anytime after 1(c) and before 7.
- 4(a) could be done earlier, anytime before 6. And 5 could be done earlier, anytime after 1(c) and before 6. In particular we could move both steps before the BPF compilation. This may improve the odds that Envoy will get configured before the timeout. But note that 6 must still be done after BPF compilation.

**Comment by rlenglet** [comment]
@joestringer I think the only thing that needs to change to support a smoother transition from L4->L7 is 1(a): we need to keep existing L3/L4 entries that will get updated in 4(b) with a non-zero proxy port.

**Comment by squeed** [comment]
I'm seeing this cause issues on a real cluster on an agent restart; this is definitely still a problem. Spurious policy drops occur between `e.syncPolicyMap()` and the end of regeneration.

**Comment by sathieu** [comment]
We today had a problem on the three control-planes of one of our clusters: during OS upgrade (after drain + upgrade + reboot + uncordon) kube-apiserver and etcd went very high on memory untill kube-vip (and others) was oom-killed.
Almost at the same time (before or after?), etcd members were unable to reach each others.

We are using Host Network Policy for ingress traffic, and the fix was to temporarily [enable the AuditMode](https://docs.cilium.io/en/latest/security/host-firewall/#enable-policy-audit-mode-for-the-host-endpoint) on node endpoints.

This looked like cilium-agent was oom-killed (which leads to this issue) but this oom-kill is not in the log. A few other process were oom-killed including cilium-operator.

Can this be related to this issue? Any other reason why cilium would block traffic in this case?

**Comment by joestringer** [comment]
Highly unlikely. This issue is about identifying one source of transient drops during normal operations and designing a solution to that very specific problem. On the other hand the problem you describe sounds like it's related to how Cilium bootstraps policy in an environment with an unreliable control plane. I think that the problem you describe is worthwhile to investigate (albeit tricky to do so), but should probably be pursued in a separate discussion.

**Comment by joestringer** [comment]
Relating to the original issue here, we discussed it a bit during the community meeting and highlighted some of the following concerns:
- Performance with map-in-map is worth evaluating. Maybe OK with an array of maps.
- This might help to improve unwinding / revert of policy steps on failures.
  - Policy is consumed by at least 3 different places - BPF, Envoy, Proxy in cilium-agent
  - Any failures in any of these can trigger the revert of policy changes

I think it would also be useful to understand the motivation for doing this atomic swapping, how it would integrate into the overall policy calculation and what we hope to achieve with this solution. That could help us to make tradeoffs or explore alternatives.
